protocols/dcutr/src/behaviour.rs

```diff
@@ -0,0 +1,401 @@
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! [`NetworkBehaviour`] to act as a direct connection upgrade through relay node.
+
+use crate::handler;
+use crate::protocol;
+use either::Either;
+use libp2p_core::connection::{ConnectedPoint, ConnectionId};
+use libp2p_core::multiaddr::Protocol;
+use libp2p_core::{Multiaddr, PeerId};
+use libp2p_swarm::dial_opts::{self, DialOpts};
+use libp2p_swarm::{
+    DialError, IntoProtocolsHandler, NetworkBehaviour, NetworkBehaviourAction, NotifyHandler,
+    PollParameters, ProtocolsHandler, ProtocolsHandlerUpgrErr,
+};
+use std::collections::{HashMap, HashSet, VecDeque};
+use std::task::{Context, Poll};
+use thiserror::Error;
+
+/// The events produced by the [`Behaviour`].
+#[derive(Debug)]
+pub enum Event {
+    InitiateDirectConnectionUpgrade {
+        remote_peer_id: PeerId,
+        local_relayed_addr: Multiaddr,
+    },
+    RemoteInitiatedDirectConnectionUpgrade {
+        remote_peer_id: PeerId,
+        remote_relayed_addr: Multiaddr,
+    },
+    DirectConnectionUpgradeSucceeded {
+        remote_peer_id: PeerId,
+    },
+    DirectConnectionUpgradeFailed {
+        remote_peer_id: PeerId,
+        error: UpgradeError,
+    },
+}
+
+#[derive(Debug, Error)]
+pub enum UpgradeError {
+    #[error("Failed to dial peer.")]
+    Dial,
+    #[error("Failed to establish substream: {0}.")]
+    Handler(ProtocolsHandlerUpgrErr<void::Void>),
+}
+
+pub struct Behaviour {
+    /// Queue of actions to return when polled.
+    queued_actions: VecDeque<Action>,
+
+    /// All direct (non-relayed) connections.
+    direct_connections: HashMap<PeerId, HashSet<ConnectionId>>,
+}
+
+impl Behaviour {
+    pub fn new() -> Self {
+        Behaviour {
+            queued_actions: Default::default(),
+            direct_connections: Default::default(),
+        }
+    }
+}
+
+impl NetworkBehaviour for Behaviour {
+    type ProtocolsHandler = handler::Prototype;
+    type OutEvent = Event;
+
+    fn new_handler(&mut self) -> Self::ProtocolsHandler {
+        handler::Prototype::UnknownConnection
+    }
+
+    fn addresses_of_peer(&mut self, _peer_id: &PeerId) -> Vec<Multiaddr> {
+        vec![]
+    }
+
+    fn inject_connected(&mut self, _peer_id: &PeerId) {}
```

Comment:
Nit: I think these have a default now.

Reply:
You are right. Removed with 831deb5625a58d3b2b65e7873d6bd0d58561aa4c.